### PR TITLE
chore: use `expect`'s own `MatcherState` for matchers

### PR DESCRIPTION
**Pull request by SimenB**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/facebook/jest/blob/895136f3f4be9155f1b60b55e77b8c285feb6a73/packages/expect/src/types.ts#L31-L53
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by SimenB** [comment]
I'm not sure what the CI failure means...

**Comment by SimenB** [comment]
This should be merged _after_ #41839 as that one actually has breaking changes. Seeing as this is failing CI, it'll probably happen anyway, I just wanna make a note of it in case something happens 🙂 

---

As for the CI error - is it because Jest depends on `@types/node`, and there's some sort of duplication going on?

File: types/jest/index.d.ts [hunk 1]
```diff
@@ -29,7 +29,7 @@
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 3.0
 
-/// <reference types="jest-diff" />
+/// <reference types="expect" />
```
Comment (ExE-Boss) [inline types/jest/index.d.ts]:
```suggestion
import { MatcherState as JestMatcherState } from "expect";
```

File: types/jest/index.d.ts [hunk 2]
```diff
@@ -456,69 +456,12 @@ declare namespace jest {
...
-    }
-
     interface ExpectExtendMap {
         [key: string]: CustomMatcher;
     }
 
-    type MatcherContext = MatcherUtils & Readonly<MatcherState>;
     type CustomMatcher = (
-        this: MatcherContext,
+        this: import("expect").MatcherState,
```
Comment (ExE-Boss) [inline types/jest/index.d.ts]:
```suggestion
        this: MatcherState,
```

File: types/jest/index.d.ts [hunk 3]
```diff
@@ -706,7 +640,7 @@ declare namespace jest {
         not: InverseAsymmetricMatchers;
 
         setState(state: object): void;
-        getState(): MatcherState & Record<string, any>;
+        getState(): import('expect').MatcherState & Record<string, any>;
```
Comment (ExE-Boss) [inline types/jest/index.d.ts]:
```suggestion
        getState(): Record<string, any>;
```

**Review by ExE-Boss — COMMENTED** [review]
This&nbsp;should&nbsp;re‑export the&nbsp;`MatcherState` from&nbsp;`"expect"`.

File: types/jest/index.d.ts [hunk 4]
```diff
@@ -456,69 +456,12 @@ declare namespace jest {
...
-        equals(a: any, b: any, customTesters?: EqualityTester[], strictCheck?: boolean): boolean;
-        [other: string]: any;
-    }
-
     interface ExpectExtendMap {
         [key: string]: CustomMatcher;
     }
 
-    type MatcherContext = MatcherUtils & Readonly<MatcherState>;
     type CustomMatcher = (
```
Comment (ExE-Boss) [inline types/jest/index.d.ts]:
```suggestion
    type MatcherState = JestMatcherState;
    type CustomMatcher = (
```

File: types/jest/index.d.ts (hunk 1 above)
Comment (SimenB) [reply types/jest/index.d.ts]: That won't work, unfortunately. See https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32891#discussion_r255284875

Would be lovely to fix it, though.

(We don't need the `references` I'll rebase this after #41852 is merged)

File: types/jest/index.d.ts (hunk 4 above)
Comment (ExE-Boss) [inline types/jest/index.d.ts]:
```suggestion
    type MatcherState = import("expect").MatcherState;
    type CustomMatcher = (
```

File: types/jest/index.d.ts (hunk 1 above)
Comment (ExE-Boss) [reply types/jest/index.d.ts]: In&nbsp;that&nbsp;case you&nbsp;can do&nbsp;<…>.

Comment (ExE-Boss) [reply types/jest/index.d.ts]: In&nbsp;that&nbsp;case you&nbsp;can do&nbsp;<…>.

Comment (SimenB) [reply types/jest/index.d.ts]: yeah, good call. done 🙂

**Comment by SimenB** [comment]
Rebased now - it'll probably still fail CI though. Anyone with a clever idea on how to fix it?

**Review by ExE-Boss — APPROVED** [review]
Well, the&nbsp;build&nbsp;failure seems&nbsp;unrelated.

**Comment by SimenB** [comment]
I poked @orta, hopefully he'll have some tricks up his sleeve 🙂

**Comment by SimenB** [comment]
Shaddap bot, waiting for @orta to take a look after 3.8 RC release

**Comment by SimenB** [comment]
A somewhat smaller version of this in #58784.

Also, Jest 28 will export way better types for `expect`, so this PR should be possible at that point

**Pull request closed without merging**